### PR TITLE
Add support for ruby 1.9 hash syntax

### PR DESCRIPTION
**Pull request by jaredbeck**
This would represent the first configuration option, AFAICT.  I made it configurable through the constructor, but I'm open to other ideas of course.

I wasn't sure how to organize the tests, sorry if they're not following the convention.

This will probably have a merge conflict with https://github.com/seattlerb/ruby2ruby/pull/36, but I can sort it out :)  Whatever order you want to merge in is fine.

### CODE REVIEW

**Comment by jaredbeck** [comment]
Hmm.  I just realized `process_hash_pair` is a bad name.  It makes it sound like there's a `:hash_pair` `sexp_type`.  Any suggestions on a better name / design?

File: lib/ruby2ruby.rb [hunk 1]
```diff
@@ -481,18 +490,27 @@ def process_hash(exp) # :nodoc:
...
-      rhs = "(#{rhs})" unless [:lit, :str].include? t # TODO: verify better!
-
-      result << "#{lhs} => #{rhs}"
+      result << process_hash_pair(lhs, rhs)
     end
 
     return result.empty? ? "{}" : "{ #{result.join(', ')} }"
   end
 
+  def process_hash_pair(lhs, rhs)
```
Comment (zenspider) [inline lib/ruby2ruby.rb]: I agree with the name being bad, tho Id on't have a good suggestion. I don' think you gain much with this refactoring. I'd be tempted to unfactor it back.

File: lib/ruby2ruby.rb [hunk 2]
```diff
@@ -60,8 +60,17 @@ class Ruby2Ruby < SexpProcessor
...
-    super
+  HASH_SYNTAXES = [:ruby18, :ruby19]
+  RUBY_19_HASH_KEY = /\A[a-z][_a-zA-Z0-9]+\Z/
+
+  ##
+  # Options:
+  #
+  # - `:hash_syntax` - either `:ruby18` or `:ruby19`
+
+  def initialize(opts = {})
```
Comment (zenspider) [inline lib/ruby2ruby.rb]: `opts` -> `option`. English is good and you only ever access one at a time.

File: lib/ruby2ruby.rb [hunk 3]
```diff
@@ -481,18 +490,27 @@ def process_hash(exp) # :nodoc:
...
 
     return result.empty? ? "{}" : "{ #{result.join(', ')} }"
   end
 
+  def process_hash_pair(lhs, rhs)
+    rhs_type = rhs.first
+    rhs = process rhs
+    rhs = "(#{rhs})" unless [:lit, :str].include? rhs_type # TODO: verify better!
+
+    if @hash_syntax == :ruby19 && hash_key_is_ruby19_compatible?(lhs)
```
Comment (zenspider) [inline lib/ruby2ruby.rb]: Use accessors, not ivars.

File: lib/ruby2ruby.rb [hunk 4]
```diff
@@ -481,18 +490,27 @@ def process_hash(exp) # :nodoc:
...
+      lhs.shift # discard type
+      "#{lhs.shift}: #{rhs}"
```
Comment (zenspider) [inline lib/ruby2ruby.rb]: I'd rather this just call process.

**Comment by zenspider** [comment]
Start with `rake SIMPLE=1` and get that to fully pass. Usually that's all you need unless there are secondary issues that only come through via self-translation (which usually means you need another test).

**Comment by zenspider** [comment]
I'm getting:

```
% rake SIMPLE=1
/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby -w -I../../minitest/dev/lib:lib:../../ruby_parser/dev/lib:../../sexp_processor/dev/lib:lib:bin:test:. -e 'require "rubygems"; gem "minitest"; require "minitest/autorun"; require "test/test_ruby2ruby.rb"' -- 


Run options: --seed 44624

# Running:

...................................................................................................................................................................................................................................................................F..................................................................F............................................................................................................................................

Finished in 0.470601s, 992.3481 runs/s, 2728.4260 assertions/s.

  1) Failure:
TestRuby2Ruby#test_process_hash_ruby19_one_pair [/Users/ryan/Work/p4/zss/src/ruby2ruby/dev/test/test_ruby2ruby.rb:110]:
Expected: "{ foo: \"bar\" }"
  Actual: "{ :foo => \"bar\" }"


  2) Failure:
TestRuby2Ruby#test_process_hash_ruby19_mixed_pairs [/Users/ryan/Work/p4/zss/src/ruby2ruby/dev/test/test_ruby2ruby.rb:125]:
--- expected
+++ actual
@@ -1 +1 @@
-"{ foo: \"bar\", 0.7 => \"baz\" }"
+"{ :foo => \"bar\", 0.7 => \"baz\" }"


467 runs, 1284 assertions, 2 failures, 0 errors, 0 skips
```

**Comment by jaredbeck** [comment]
Hmm, I'm not sure why you got those test failures.  I've rebased and `rake` passes for me, without `SIMPLE`.

Two topics worth discussing:

1. Should we wrap the `option` hash in a configuration object?  One advantage would be to provide a home for the proposed `extract_option` method.
2. Given the test suite's `morph_and_eval` e.g. `TestRuby2Ruby2`, how can I use `assert_raises(InvalidOption)` in my tests, instead of the `begin/rescue` workaround I'm currently using?

Thanks!

**Comment by jaredbeck** [comment]
I've scaled this back a bit.  It no longer converts string keys to symbol keys because that would not be faithful to the input.

**Comment by zenspider** [comment]
Thanks, but I think I'm gonna have to officially say "no, thank you" to this. It's a lot of complexity for dubious payoff (to me). Since it can be arranged as a simple override of `process_hash` in a subclass, it probably belongs there. If you want help packaging this up in its own gem, I'm happy to provide assistance.

**Comment by jaredbeck** [comment]
> Thanks, but I think I'm gonna have to officially say "no, thank you" to this.

No worries. I started a fork in April '15 called [sexp2ruby](https://github.com/jaredbeck/sexp2ruby) that had "style" as a goal, and it served it's purpose well, but lately I've been really impressed with rubocop's `--autocorrect` feature and in light of that am not sure I want to maintain sexp2ruby anymore. :)

**Pull request closed without merging**